### ISSUE TITLE
C/CXX Bazel flags are propagated to crate dependencies

### ISSUE THREAD

**Issue by fmorency** [bazelbuild/rules_rust#1550]
Follow-up from a thread on the Bazel Slack. CC @illicitonion.

**MRE**
Clone MRE in https://github.com/fmorency/bazelsyscov and run
```
$ CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index
$ bazel coverage --combined_report=lcov //...:all-targets
```
The above will fail to build.

Running the following works as expected

```
$ bazel build //...:all-targets
```

Tested using Bazel 5.2

**Problem**
Bazel is propagating C/CXX flags to crate dependencies. I stumbled on at least one crate where this is causing an issue when using `bazel coverage`. 

Bazel propagates the `--coverage` flag to the GCC build of the `-sys` crate resulting in a compilation error due to additional missing flags according to the `-sys` crate's Makefile.

**Workaround**
Setting
```
[crate.annotation(
        build_script_env = {
            "CRATE_CC_NO_DEFAULTS": "",
            "CFLAGS": "",
            "CXXFLAGS": "",
            "LDFLAGS": "",
        },
    )]
```
on the problematic crate.

**Comment by UebelAndre** [comment]
Can you link an example of the error?

**Comment by fmorency** [comment]
@UebelAndre @illicitonion

Here's the error I'm getting when trying to build the MRE as-is. I will work on creating a Docker container.

```
...
running: "/usr/bin/gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "/home/fmorency/.cache/bazel/_bazel_fmorency/9d3838d77d1843210fc908b53c516661/sandbox/linux-sandbox/93/execroot/__main__/bazel-out/k8-fastbuild/bin/external/crate_index__librocksdb-sys-0.8.0-7.4.4/librocksdb-sys_build_script.out_dir/rocksdb/db/memtable.o" "-c" "rocksdb/db/memtable.cc"
cargo:warning=In file included from rocksdb/db/memtable.h:25,
cargo:warning=                 from rocksdb/db/memtable.cc:10:
cargo:warning=rocksdb/memory/concurrent_arena.h: In function 'size_t& _ZTWN7rocksdb15ConcurrentArena9tls_cpuidE()':
cargo:warning=rocksdb/memory/concurrent_arena.h:100:30: error: function starts on a higher line number than it ends [-Werror=coverage-invalid-line-number]
cargo:warning=  100 |   static thread_local size_t tls_cpuid;
cargo:warning=      |                              ^~~~~~~~~
cargo:warning=cc1plus: some warnings being treated as errors
exit status: 1
running: "/usr/bin/gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "/home/fmorency/.cache/bazel/_bazel_fmorency/9d3838d77d1843210fc908b53c516661/sandbox/linux-sandbox/93/execroot/__main__/bazel-out/k8-fastbuild/bin/external/crate_index__librocksdb-sys-0.8.0-7.4.4/librocksdb-sys_build_script.out_dir/rocksdb/db/memtable_list.o" "-c" "rocksdb/db/memtable_list.cc"
exit status: 0

--stderr:


error occurred: Command "/usr/bin/gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "/home/fmorency/.cache/bazel/_bazel_fmorency/9d3838d77d1843210fc908b53c516661/sandbox/linux-sandbox/93/execroot/__main__/bazel-out/k8-fastbuild/bin/external/crate_index__librocksdb-sys-0.8.0-7.4.4/librocksdb-sys_build_script.out_dir/rocksdb/db/memtable.o" "-c" "rocksdb/db/memtable.cc" with args "gcc" did not execute successfully (status code exit status: 1).



Target //:bazelsyscov failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 234.837s, Critical Path: 228.62s
INFO: 215 processes: 123 internal, 92 linux-sandbox.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Comment by fmorency** [comment]
@illicitonion 

Dockerfile
```dockerfile
FROM archlinux:latest
RUN pacman -Syu && pacman -S --noconfirm bazel git gcc gcc-libs clang llvm
RUN git clone https://github.com/fmorency/bazelsyscov.git
WORKDIR bazelsyscov
RUN CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index
RUN bazel coverage --combined_report=lcov //...:all-targets
```

Save this Dockerfile somewhere and run `docker build . -t bazelsyscov:latest` in the folder where the Dockerfile is located.

You should get
```
cargo:warning=In file included from rocksdb/db/memtable.h:25,
cargo:warning=                 from rocksdb/db/memtable.cc:10:
cargo:warning=rocksdb/memory/concurrent_arena.h: In function 'size_t& _ZTWN7rocksdb15ConcurrentArena9tls_cpuidE()':
cargo:warning=rocksdb/memory/concurrent_arena.h:100:30: error: function starts on a higher line number than it ends [-Werror=coverage-invalid-line-number]
cargo:warning=  100 |   static thread_local size_t tls_cpuid;
cargo:warning=      |                              ^~~~~~~~~
cargo:warning=cc1plus: some warnings being treated as errors
exit status: 1
running: "/usr/sbin/gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "/root/.cache/bazel/_bazel_root/2e195946f5604c5b57437e2c98b22d8b/sandbox/processwrapper-sandbox/93/execroot/__main__/bazel-out/k8-fastbuild/bin/external/crate_index__librocksdb-sys-0.8.0-7.4.4/librocksdb-sys_build_script.out_dir/rocksdb/db/memtable_list.o" "-c" "rocksdb/db/memtable_list.cc"
exit status: 0

--stderr:
Failed to run rustfmt: No such file or directory (os error 2) (non-fatal, continuing)


error occurred: Command "/usr/sbin/gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "rocksdb/include/" "-I" "rocksdb/" "-I" "rocksdb/third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "/root/.cache/bazel/_bazel_root/2e195946f5604c5b57437e2c98b22d8b/sandbox/processwrapper-sandbox/93/execroot/__main__/bazel-out/k8-fastbuild/bin/external/crate_index__librocksdb-sys-0.8.0-7.4.4/librocksdb-sys_build_script.out_dir/rocksdb/db/memtable.o" "-c" "rocksdb/db/memtable.cc" with args "gcc" did not execute successfully (status code exit status: 1).



Target //:bazelsyscov failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 234.311s, Critical Path: 224.81s
INFO: 215 processes: 123 internal, 92 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
ERROR: No test targets were found, yet testing was requested
FAILED: Build did NOT complete successfully
```

**Comment by illicitonion** [comment]
I did a little debugging of this - it looks like this is some bug in the interaction between thread-local storage and gcc's coverage mode.

I'll try to debug a little further and file a bug upstream, but in the mean time, I'd also say it's a bug that CFLAGS/CXXFLAGS _override_ rather than _append to_ the cc_toolchain-supplied values, so I'll spin up a PR which fixes that too. Then to work around the bug you'd set an annotation which looks more like `build_script_env = {"CXXFLAGS": "-Wno-error=coverage-invalid-line-number"}`, to disable this particular bug's error, rather than wiping out CXXFLAGS completely.

**Comment by illicitonion** [comment]
(Also, thanks for providing the docker image repro, it was really helpful!)

**Comment by fmorency** [comment]
Thanks for confirming the issue and looking into it.

**Comment by illicitonion** [comment]
I spent a while trying to minimise the bug in gcc, but didn't get _super_ far. I suspect it's some interaction between templates, closures, and thread-locals when using code coverage (which... Kind of makes sense as difficult features to interact).

@fmorency Unfortunately I don't currently have permission to file a bug against gcc, but I'd encourage you to do so here https://gcc.gnu.org/bugzilla/ - your core repro issues are:
1. Using gcc 12.2.0 - gcc 11 doesn't repro
2. You're building https://github.com/facebook/rocksdb/releases/tag/v7.4.4 (but they don't care about this) and your complete command line is: `"gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-U_FORTIFY_SOURCE" "-fstack-protector" "-Wall" "-Wunused-but-set-parameter" "-Wno-free-nonheap-object" "-fno-omit-frame-pointer" "-std=c++0x" "--coverage" "-fno-canonical-system-headers" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "include/" "-I" "." "-I" "./third-party/gtest-1.8.1/fused-src/" "-I" "." "-std=c++17" "-Wsign-compare" "-Wshadow" "-Wno-unused-parameter" "-Wno-unused-variable" "-Woverloaded-virtual" "-Wnon-virtual-dtor" "-Wno-missing-field-initializers" "-Wno-strict-aliasing" "-Wno-invalid-offsetof" "-DNDEBUG=1" "-DOS_LINUX" "-DROCKSDB_PLATFORM_POSIX" "-DROCKSDB_LIB_IO_POSIX" "-DROCKSDB_SUPPORT_THREAD_LOCAL" "-DHAVE_UINT128_EXTENSION=1" "-o" "db/memtable.o" "-c" "db/memtable.cc"`
3. You can create a `.ii` file they'll want by running the above command line with `-v -save-temps` added as args directly after `gcc`

In the mean time, once https://github.com/bazelbuild/rules_rust/pull/1663 is merged I'll put up a PR allowing appending rather than overriding env vars.

**Comment by fmorency** [comment]
@illicitonion I think it's important to mention that the project builds fine with Cargo and code coverage instrumentation enabled.

I would (naively) expect Bazel to be able to reproduce the same behavior; the Bazel workspace relies on the Cargo manifest to build its index.

Bazel is doing something that Cargo isn't.

**Comment by illicitonion** [comment]
It's a good thing to note - could you turn up the logging of the cargo version (at least running `cargo build -vv`) to see if you can get to the exact `gcc` invocation you're seeing with Cargo? (I don't actually know how to run a cargo build withwith coverage!)

I strongly suspect the difference comes from `cargo` using a slightly different set of flags than the default Bazel C++ toolchain does, and being able to compare the exact command lines would tell us exactly what.

Also, are you sure that rocksdb ends up getting instrumented for coverage when building with cargo? I can believe that the `cargo` build script for rocksdb _isn't_ propagating the request for coverage to gcc, and so the rocksdb symbols _don't_ end up getting coverage data.

**Comment by illicitonion** [comment]
The flag merging instead of overriding change is here: https://github.com/bazelbuild/rules_rust/pull/1675

**Comment by fmorency** [comment]
Closing as fixed by #1675.
